lib/imparse.c

```diff
@@ -216,6 +217,92 @@ EXPORTED int imparse_isnumber(const char *s)
     return 1;
 }
 
+static int reject_http_method_tag(const char *s)
+{
+    /* Don't like tags that match HTTP methods that accept a request body!
+     * Keep this up to date with http_methods[] in httpd.c
+     * and test_istag() in imparse.testc
+     */
+    switch (s[0]) {
+    case 'A':
```

Comment:
httpd.c only lists methods that it supports.  It isn't an exhaustive list.  A complete list can be found here:
https://www.iana.org/assignments/http-methods/http-methods.xhtml#methods.  I know that all of the methods in RFC 3253 accept bodies.

I wonder if, rather than using a switch statement, we could create a strarray via strarray_split() in service_init() and then just use strarray_find() to check if a tag is an HTTP method.

Reply:
strarray_find would call strcmp for every string in the array until it finds it, or has checked them all -- which would be similar to an if chain, except that strarray also by necessity chases pointers around, so it can't benefit from prefetching if its strings aren't already in cache.  We could perhaps improve that by pre-sorting the strarray (or the hypothetical if chain) by expected frequency, such that at least we'd be comparing the most likely strings first, but that seems like a lot of work.

The switch approach only calls strcmp 5 times in the worst case (if the tag starts with 'P' and isn't one of the first four of those), and 0 in the likely case (when the tag doesn't start with any of these letters), and when it does call strcmp, the comparators are accessed directly with no pointer chasing.

It might not make much difference in practice, though.  If this were unusual or tricky code, I'd be in favour of simplifying it for readability/maintainability, at maybe some performance loss.  But switching on the first character is an idiom we already use a bunch in Cyrus, it's not unusual, nor hard to maintain.